### PR TITLE
Multi subnets support

### PR DESCRIPTION
**Pull request by FinchPowers**
StarCluster in VPC mode only supported a single VPC.
We want many to allow us to bid in the cheapest zone.

### CODE REVIEW

File: starcluster/awsutils.py [hunk 1]
```diff
@@ -1613,19 +1613,27 @@ def show_console_output(self, instance_id):
...
+        Find cheapest zone.
+
+        zone_filter, if a list, zones to consider, if None, all zones
+            considered
+        """
         min_price = 9999
         min_zone = None
-        for z in ['a', 'c', 'd']:
-            zone = 'us-east-1' + z
+        for zone in self.conn.get_all_zones():
```
Comment (FinchPowers) [inline starcluster/awsutils.py:1625]: This is not related to the current PR. Removes the hardcoded zone list and fetches the list from ec2 instead.

File: starcluster/cluster.py [hunk 2]
```diff
@@ -423,7 +424,7 @@ def __init__(self,
                  cluster_group=None,
                  force_spot_master=False,
                  disable_cloudinit=False,
-                 subnet_id=None,
+                 subnet_ids=[],
```
Comment (FinchPowers) [inline starcluster/cluster.py:427]: All the way through this file, subnet_id is now subnet_ids. It disconnects us from the core version.

File: starcluster/cluster.py [hunk 3]
```diff
@@ -665,11 +667,15 @@ def __getstate__(self):
     def _security_group(self):
         return static.SECURITY_GROUP_TEMPLATE % self.cluster_tag
 
+    def get_vpc_id(self, subnet_id):
+        subnet = self.ec2.get_subnet(self.subnet_ids[0])
+        return subnet.vpc_id
+
     @property
-    def subnet(self):
```
Comment (FinchPowers) [inline starcluster/cluster.py:669]: This made sense before because there was a single subnet. Not anymore.

File: starcluster/cluster.py [hunk 4]
```diff
@@ -678,19 +684,27 @@ def cluster_group(self):
         sg = self.ec2.get_group_or_none(self._security_group)
         if not sg:
             desc = 'StarCluster-%s' % static.VERSION.replace('.', '_')
-            if self.subnet:
+            if self.subnet_ids:
```
Comment (FinchPowers) [inline starcluster/cluster.py:687]: Previously, the check for "is this a vpc?" was made via "if self.subnet". I've replaced it with "if self.subnet_ids".
Maybe we should consider a function or something more self-explaining.

File: starcluster/cluster.py [hunk 5]
```diff
@@ -2508,44 +2543,75 @@ def validate_userdata(self):
                 "to store internal metadata" % ud_size_kb)
 
     def validate_vpc(self):
```
Comment (FinchPowers) [inline starcluster/cluster.py:2545]: Bunch of changes to test all subnets.

File: starcluster/cluster.py [hunk 6]
```diff
@@ -2508,44 +2543,75 @@ def validate_userdata(self):
...
+                    self.cluster.vpc_id)
+            # for subnet_id in self.cluster.subnet_ids:
+            #     rtables = self.cluster.ec2.get_route_tables(filters={
+            #         'association.subnet-id': subnet_id,
+            #         'route.destination-cidr-block': static.WORLD_CIDRIP,
+            #         'route.gateway-id': gws[0].id})
+            #     if not rtables:
+            #         raise exception.ClusterValidationError(
+            #             "No route to %s found for subnet: %s" %
+            #             (static.WORLD_CIDRIP, subnet_id))
```
Comment (FinchPowers) [inline starcluster/cluster.py:2611]: @jraby told me this test is somewhat invalid and that failing to comply doesn't necessarily mean that the cluster won't work.

File: starcluster/cluster.py [hunk 7]
```diff
@@ -2508,44 +2543,75 @@ def validate_userdata(self):
...
-        if self.cluster.subnet_id:
-            try:
-                assert self.cluster.subnet is not None
-            except exception.SubnetDoesNotExist as e:
-                raise exception.ClusterValidationError(e)
-            azone = self.cluster.availability_zone
-            szone = self.cluster.subnet.availability_zone
+        if not self.cluster.subnet_ids:
+            return
+        if self.cluster.public_ips is False:
```
Comment (wsourdeau) [inline starcluster/cluster.py:2548]: Ca me semble suspect. C'est pas "None" que tu devrais tester?

Comment (FinchPowers) [inline starcluster/cluster.py:2548]: C'est du code du core qui a changé d'indentation. J'ai gardé la même condition.

File: starcluster/cluster.py (hunk 6 above)
Comment (jraby) [inline starcluster/cluster.py:2611]: This code checks if the subnet has an route table  **explicitly associated**  to it and if that route table has a route for 0.0.0.0

This is wrong since a subnet will use the 'main' route table by default and that implicit association cannot be matched by `get_route_tables()` filters. Thus, even if the subnet can reach 0.0.0.0/0 via the default route table, the check would fail.

We decided to remove the check since it was not easy to fix properly and because it works fine in our environment.  Upstream may disagree.

File: starcluster/cluster.py (hunk 7 above)
Comment (wsourdeau) [inline starcluster/cluster.py:2548]: Ok. Note que ca me semble suspect quand-meme. parce que public_ips semblent etre soit None, soit une liste.

**Comment by wsourdeau** [comment]
+1 pour moi

File: starcluster/cluster.py (hunk 7 above)
Comment (FinchPowers) [inline starcluster/cluster.py:2548]: Juste pour fin de suivi, voici comment ça peut être false.
https://github.com/datacratic/StarCluster/blob/5d9628064e695a26bebc338f84caab339f241969/starcluster/commands/start.py#L96

**Comment by FinchPowers** [comment]
Squash merged.
